tests/functional/tests/bundles/overlays/local-charm-overlay.yaml.j2

```diff
@@ -1,5 +1,5 @@
 applications:
   ud-ldap-server:
-    charm: "{{ CHARM_LOCATION }}/{{ charm_name }}.charm"
+    charm: "{{ CHARM_DIR }}/userdir-ldap_ubuntu-22.04-amd64-arm64_ubuntu-20.04-amd64-arm64.charm"
```

Comment:
Can we do this as argument? What about removing `CHARM_DIR` and hard-coded name, and use `{{ CHARM }}`, which will represent path to the charm file.
Like that, I can provide any path I want, e.g. I download specific revision or build it in different directory.

Another suggestion is that the at least CI should find out the path for charm file themselves, so we do not need to come and change this when we update base or another architecture. WDYT?

Reply:
Ah this would be nice.  I'll see what I can do. :)

Reply:
It looks like there are a couple of routes we could go here. Here are some suggestions for consideration:

# 1

If we want the workflow to automatically pick up the charm file, we can parse the output from `charmcraft pack`:

```
# single charm output (only stdout shown)
$ charmcraft -v pack --format json
{
    "charms": [
        "userdir-ldap_ubuntu-22.04-amd64-arm64_ubuntu-20.04-amd64-arm64.charm"
    ]
}

# multiple charms output (only stdout shown)
$ charmcraft -v pack --format json
{
    "charms": [
        "userdir-ldap_ubuntu-22.04-amd64-arm64.charm",
        "userdir-ldap_ubuntu-20.04-amd64-arm64.charm"
    ]
}
```

(I showed both outputs because we will likely have cases soon where charmcraft will build multiple charms.)
Then we could run the functional tests against all the output charms.  However, this presents a problem where the workflow gets more complex, and we need to infer which series to run the tests with (based on the charm names).

# 2

We could have more control and bake all the combinations into the test matrix, kind of like:

```
    strategy:
      fail-fast: false
      matrix:
        test-command: ['tox -e func']
        juju-channel: ["3.4/stable"]
        include:
          - runs-on: [ubuntu-22.04]
            charm-file: 'userdir-ldap_ubuntu-22.04-amd64.charm'
            series: jammy
          - runs-on: [ubuntu-22.04]
            charm-file: 'userdir-ldap_ubuntu-20.04-amd64.charm'
            series: focal
          - runs-on: [Ubuntu_ARM64_4C_16G_01]
            charm-file: 'userdir-ldap_ubuntu-22.04-arm64.charm'
            series: jammy
          - runs-on: [Ubuntu_ARM64_4C_16G_01]
            charm-file: 'userdir-ldap_ubuntu-20.04-arm64.charm'
            series: focal
```

I think this is what I'm leaning toward, as it's clear what combinations are being tested, and each job will do a single test run.

Others?

Reply:
@rgildein I'll resolve this for now, because I've updated it to follow your suggestion.  We may need more refactoring as we consider how best to port this to automation and to apply it to other projects. :)